### PR TITLE
Add types supporting C# 8.0 Ranges.

### PR DESCRIPTION
**Pull request by baulig**

### CODE REVIEW

**Comment by baulig** [comment]
Requires https://github.com/mono/corefx/pull/250.

**Comment by baulig** [comment]
Getting native crash with this: https://github.com/mono/mono/issues/12842.

**Comment by baulig** [comment]
Adding don't merge because of the runtime crash.

**Comment by baulig** [comment]
Runtime crash has been confirmed as being caused by this change:

```
	static char Test (ReadOnlySpan<char> span, Index index)
	{
		return span[index];
	}
```

Full details in https://github.com/mono/mono/issues/12842.

**Comment by baulig** [comment]
The runtime issue has been resolved thanks to @lewurm :-)

**Comment by marek-safar** [comment]
@baulig please resolve the conflicts

File: mcs/class/corlib/corlib.dll.sources [hunk 1]
```diff
@@ -114,6 +115,8 @@ System/NumberFormatter.cs
 System/Object.cs
 ../../../external/corefx/src/System.Runtime.Extensions/src/System/OperatingSystem.cs
 System/PlatformID.cs
+../../../external/corefx/src/Common/src/CoreLib/System/Range.cs
+../../../external/corefx/src/Common/src/CoreLib/System/ReadOnlySpan.Fast.cs
 ../../../external/corefx/src/Common/src/CoreLib/System/ResolveEventArgs.cs
```
Comment (EgorBo) [inline mcs/class/corlib/corlib.dll.sources]: Don't we already have `ReadOnlySpan.Fast.cs`  and `Span.Fast.cs` in this file?

**Comment by baulig** [comment]
@monojenkins build failed

File: mcs/class/corlib/corlib.dll.sources (hunk 1 above)
Comment (baulig) [reply mcs/class/corlib/corlib.dll.sources]: We did; thanks for spotting this, I fixed it.

**Comment by baulig** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@monojenkins commit csproj and apidiff

**Pull request merged**